File: server/services/AQIService.ts
import { BaseService } from "./BaseService";
import { saveToFirestore, logStructured } from "../gcp";
import fs from "fs";
import path from "path";

export interface AQIRecord {
  timestamp: string;
  location: string;
  aqi: number;
  pm2_5: number;
  pm10: number;
  no2: number;
  co: number;
  o3: number;
  risk_status: string;
}

export class AQIService extends BaseService {
  private static instance: AQIService;

  public static getInstance(): AQIService {
    if (!AQIService.instance) {
      AQIService.instance = new AQIService();
    }
    return AQIService.instance;
  }

  private getRiskStatus(aqi: number): string {
    if (aqi <= 50) return "Good";
    if (aqi <= 100) return "Moderate";
    if (aqi <= 150) return "Unhealthy for Sensitive Groups";
    if (aqi <= 200) return "Unhealthy";
    if (aqi <= 300) return "Very Unhealthy";
    return "Hazardous";
  }

  public async fetchAQI(lat: number, lng: number, locationName = "Downtown"): Promise<AQIRecord> {
    const cacheKey = `aqi_${lat.toFixed(4)}_${lng.toFixed(4)}_${locationName}`;
    const cached = this.getCached<AQIRecord>(cacheKey);
    if (cached) return cached;

    // Use Open-Meteo Air Quality API
    const url = `https://air-quality-api.open-meteo.com/v1/air-quality?latitude=${lat}&longitude=${lng}&current=us_aqi,pm10,pm2_5,nitrogen_dioxide,carbon_monoxide,ozone`;

    try {
      const data = await this.fetchWithRetry(url);
      const current = data.current;

      const aqiVal = Math.round(current.us_aqi || 42);

      const record: AQIRecord = {
        timestamp: new Date().toISOString(),
        location: locationName,
        aqi: aqiVal,
        pm2_5: Number((current.pm2_5 || 11.2).toFixed(1)),
        pm10: Number((current.pm10 || 22.4).toFixed(1)),
        no2: Number((current.nitrogen_dioxide || 15.6).toFixed(1)),
        co: Number(((current.carbon_monoxide || 350) / 1000).toFixed(2)), // convert microg to mg/m3 approx
        o3: Number((current.ozone || 32.1).toFixed(1)),
        risk_status: this.getRiskStatus(aqiVal),
      };

      // Cache for 15 minutes
      this.setCached(cacheKey, record, 15 * 60 * 1000);

      // Save historical state
      await this.saveHistory(record, lat, lng);

      return record;
    } catch (err: any) {
      logStructured("ERROR", `Failed to fetch AQI for ${locationName} from Open-Meteo API. Using fallback.`, { error: err.message });
      // Fallback
      return {
        timestamp: new Date().toISOString(),
        location: locationName,
        aqi: 55,
        pm2_5: 14.5,
        pm10: 28.0,
        no2: 12.0,
        co: 0.45,
        o3: 35.0,
        risk_status: "Moderate",
      };
    }
  }

  private async saveHistory(data: AQIRecord, lat: number, lng: number): Promise<void> {
    const docId = `aqi_${Date.now()}_${data.location.replace(/\s+/g, "_")}`;
    const payload = { ...data, latitude: lat, longitude: lng };
    await saveToFirestore("aqi_history", docId, payload);

    // Persist locally in workspace
    try {
      const dirPath = path.join(process.cwd(), "src", "data", "history");
      if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
      }
      const historyFile = path.join(dirPath, "aqi_history.json");
      let list: any[] = [];
      if (fs.existsSync(historyFile)) {
        try {
          list = JSON.parse(fs.readFileSync(historyFile, "utf-8"));
        } catch {
          list = [];
        }
      }
      list.push(payload);
      if (list.length > 100) list.shift();
      fs.writeFileSync(historyFile, JSON.stringify(list, null, 2), "utf-8");
    } catch (e: any) {
      console.warn("Local history saving failed for AQI", e.message);
    }
  }
}

File: backend/app/services/aqi_service.py
import httpx
import logging
from typing import Dict, Any

logger = logging.getLogger("CommunityOS.AQIService")

AQI_API_URL = "https://air-quality-api.open-meteo.com/v1/air-quality"

async def fetch_live_aqi(lat: float = 40.7128, lon: float = -74.0060) -> Dict[str, Any]:
    """
    Fetches real-time air quality metrics from Open-Meteo Air Quality API.
    """
    logger.info(f"Fetching live AQI for coordinates: {lat}, {lon}...")
    params = {
        "latitude": lat,
        "longitude": lon,
        "current": "european_aqi,pm2_5,pm10,carbon_monoxide,nitrogen_dioxide,ozone"
    }
    
    for attempt in range(3):
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(AQI_API_URL, params=params)
                if response.status_code == 200:
                    data = response.json()
                    current = data.get("current", {})
                    
                    # Open-Meteo European AQI maps nicely to standard scales (0-100+)
                    aqi_val = int(current.get("european_aqi", 45))
                    pm25 = float(current.get("pm2_5", 12.0))
                    pm10 = float(current.get("pm10", 22.0))
                    no2 = float(current.get("nitrogen_dioxide", 15.0))
                    co = float(current.get("carbon_monoxide", 250.0))
                    o3 = float(current.get("ozone", 65.0))
                    
                    risk_status = "Good"
                    if aqi_val > 100:
                        risk_status = "Unhealthy"
                    elif aqi_val > 50:
                        risk_status = "Moderate"
                        
                    return {
                        "location": "Downtown",
                        "aqi": aqi_val,
                        "pm2_5": pm25,
                        "pm10": pm10,
                        "no2": no2,
                        "co": co,
                        "o3": o3,
                        "risk_status": risk_status
                    }
        except Exception as e:
            logger.warn(f"AQI API attempt {attempt + 1} failed: {e}")
            
    raise RuntimeError("Failed to fetch AQI from external Open-Meteo endpoint after 3 attempts.")
